capstone/src/main/java/com/google/sps/AssociationDatastore.java

```diff
@@ -0,0 +1,104 @@
+package com.google.sps;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.sps.data.MapData;
+import java.util.ArrayList;
+import java.util.Optional;
+
+/** Does all datastore operations related to associations */
+public class AssociationDatastore {
+
+  public static final String SURVEY_ENTITY_KIND = "Response";
+  public static final String COMMENT_PROPERTY = "text";
+  public static final String ZIPCODE = "zipCode";
+
+  private MapData mapData = new MapData();
+  private DatastoreService datastore;
+
+  /** @param datastore the datastore client */
+  public AssociationDatastore(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  /**
+   * Get the survey comment text from datastore
+   *
+   * @return an arraylist of the comment text
+   */
+  public ArrayList<AssociationInput> getComments() {
+    Query query = new Query(SURVEY_ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<AssociationInput> comments = new ArrayList<AssociationInput>();
+    for (Entity e : results.asIterable()) {
+      if (e.getProperty("association-processed") == null
+          || !((boolean) e.getProperty("association-processed"))) {
+        String message = (String) e.getProperty(COMMENT_PROPERTY);
+        try {
+          ArrayList<String> scope = mapData.getScope((String) e.getProperty(ZIPCODE));
+          comments.add(new AssociationInput(message, scope));
+          e.setProperty("association-processed", true);
+          datastore.put(e);
+        } catch (NullPointerException exception) {
+          System.err.println("Invalid zip code in response: " + (String) e.getProperty(ZIPCODE));
+        }
+      }
+    }
+    return comments;
+  }
+
+  /**
+   * Loads all previous association results from datastore
+   *
+   * @param scope the scope to load results from
+   * @return an arraylist of previous responses
+   */
+  public ArrayList<AssociationResult> loadPreviousResults(String scope) {
+    Query query = new Query(AssociationResult.ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<AssociationResult> prev = new ArrayList<AssociationResult>();
+    for (Entity entity : results.asIterable()) {
+      if (!scope.equals((String) entity.getProperty("scope"))) continue;
+      String content = (String) entity.getProperty("name");
+      float weight = (float) (double) entity.getProperty("weight");
+      float score = (float) (double) entity.getProperty("score");
```

Comment:
Do you need to cast this twice?

Reply:
Yes, if you don't explicitly do the double cast it complains that the object stored is not a float.